content/blog/2018-08-28-birds-ocr.md

```diff
@@ -0,0 +1,411 @@
+    ##   <chr>                <chr> <chr> <fs::path>                        
+    ## 1 climacteris picumnus <NA>  la    birds/n115_w1150_42399797481_o.jpg
+    ## 2 brown tree creeper   en    <NA>  birds/n115_w1150_42399797481_o.jpg
+
+So we see that the image transformation was quite useful.
+
+Now, these names look fine, but how to be sure they’re actually
+taxonomic names?
+
+#### Taxonomic name resolution
+
+The [`taxize` package](https://github.com/ropensci/taxize/) by Scott
+Chamberlain, taxonomic toolbelt for R, provides access to many fantastic
+data sources and tools for taxonomy. One of them, the Global Name
+Resolver, provides, well, resolution of taxonomic names, sadly not
+common names. `taxize::gnr_resolve` has many options, we only make use
+of the one returning only the best match from the different data
+sources.
+
+``` r
+latin <- results2$text[results2$cld2 == "la"|
+                         results2$cld3 == "la"]
+taxize::gnr_resolve(latin,
+  best_match_only = TRUE)
+```
+
+    ## # A tibble: 1 x 5
+    ##   user_supplied_name submitted_name  matched_name   data_source_tit~ score
+    ## * <chr>              <chr>           <chr>          <chr>            <dbl>
+    ## 1 climacteris picum~ Climacteris pi~ Climacteris p~ NCBI             0.988
+
+English names could be cleaned up a bit using the [`spelling`
+package](https://github.com/ropensci/spelling), also an rOpenSci’s
+package, since `spelling::spell_check_text` would output possible typos.
+
+### OCR-bird naming workflow in action!
+
+We apply the first two steps, image processing and OCR, to all images.
+
+``` r
+bird_names <- purrr::map(filenames, crop_bird) %>%
+  purrr::map_df(get_names)
+```
+
+Out of 51 images only 17 are present in the final table with possible
+names which is a bit disheartening, but we could surely do better in the
+image processing and OCR tweaking steps! Maybe one could frame the
+parameter search [as a machine learning
+problem](https://twitter.com/dmi3k/status/1024919690768990209). We
+should also keep in mind that such Natural History images are quite hard
+to parse.
+
+The name resolution information can be added to this table.
+
+``` r
+safe_resolve <- function(text){
+  
+  results <- taxize::gnr_resolve(text,
+                                 best_match_only = TRUE)
+  
+  if(nrow(results) == 0){
+    list(NULL)
+  }else{
+    list(results)
+  }
+}
+
+bird_names <- dplyr::group_by(bird_names, text) %>%
+  dplyr::mutate(gnr = ifelse(cld2 == "la" | cld3 == "la",
+                             safe_resolve(text),
+                             list(NULL)))
+```
+
+We do not get much resolution, but we knew the names weren’t very good
+to start with. A better idea here might be to get a full list of names
+of Australian birds, potentially leveraging [the `taxizedb` package by
+Scott Chamberlain](https://github.com/ropensci/taxizedb), and to then
+fuzzy-match them with the names we have.
```

Comment:
you should comment on whether that was helpful or not

Reply:
I can't because it's just a random idea I wrote. Will add "untested".